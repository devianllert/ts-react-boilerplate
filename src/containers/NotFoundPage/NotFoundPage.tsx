import React, { ReactElement } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';

import Button from '../../components/Button';

import styles from './NotFoundPage.module.scss';

const NotFoundPage = (): ReactElement => {
  const { t } = useTranslation();

  return (
    <div className={styles.content}>
      <Helmet>
        <title>{t('notFound.headTitle')}</title>
      </Helmet>

      <h1 className={styles.logo}>404</h1>
      <h2 className={styles.title}>{t('notFound.title')}</h2>
      <span className={styles.subtitle}>{t('notFound.subtitle')}</span>

      <Button to="/">{t('notFound.button')}</Button>
    </div>
  );
};

export default NotFoundPage;
