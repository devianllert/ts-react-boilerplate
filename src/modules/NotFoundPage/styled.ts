import styled from 'styled-components';

export const Image = styled.img`
  width: 100%;
  max-height: 400px;
  margin-bottom: 16px;
`;

export const CenteredBox = styled.div`
  min-height: 100vh;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
