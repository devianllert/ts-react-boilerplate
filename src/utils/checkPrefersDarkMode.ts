/**
 * Check browser or OS preference for dark mode
 */

const checkPrefersDarkMode = (): boolean => {
  const preferDarkQuery = '(prefers-color-scheme: dark)';

  const mediaQuery = matchMedia(preferDarkQuery);

  return mediaQuery.media === preferDarkQuery;
};

export default checkPrefersDarkMode;
