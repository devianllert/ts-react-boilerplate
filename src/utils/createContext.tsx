import React, {
  createContext,
  useState,
  useContext,
  Provider,
  Dispatch,
  SetStateAction,
  PropsWithChildren,
  ReactNode,
  Context,
} from 'react';

type UpdateType<A> = Dispatch<SetStateAction<A>>;

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export const createCtx = <A extends {} | null>(): readonly [() => A, Provider<A | undefined>] => {
  const ctx = createContext<A | undefined>(undefined);

  const useCtx = (): A => {
    const c = useContext(ctx);

    if (c === undefined) {
      throw new Error('useCtx must be inside a Provider with a value');
    }

    return c;
  };

  return [useCtx, ctx.Provider] as const;
};

export const createCtxWithUpdate = <A extends any>(
  defaultValue: A,
): readonly [
  Context<{ state: A; update: Dispatch<SetStateAction<A>> }>,
  (props: { children?: ReactNode }) => ReactNode,
] => {
  const defaultUpdate: UpdateType<typeof defaultValue> = () => defaultValue;

  const ctx = createContext({
    state: defaultValue,
    update: defaultUpdate,
  });

  const ContextProvider = (props: PropsWithChildren<{}>): ReactNode => {
    const [state, update] = useState(defaultValue);
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <ctx.Provider value={{ state, update }} {...props} />;
  };

  return [ctx, ContextProvider] as const;
};
