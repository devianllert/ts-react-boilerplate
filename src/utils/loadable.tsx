/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/jsx-props-no-spreading */

import React, {
  lazy,
  Suspense,
  ReactElement,
  ComponentType,
} from 'react';

export type ImportComponent = Promise<{ default: any }>;

const loadable = (importFunc: () => ImportComponent, { fallback = null }): ComponentType => {
  const LazyComponent: ComponentType = lazy(importFunc);

  return (props: object): ReactElement => (
    <Suspense fallback={fallback}>
      <LazyComponent {...props} />
    </Suspense>
  );
};

export default loadable;
