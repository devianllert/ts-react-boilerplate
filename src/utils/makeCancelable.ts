/* eslint-disable prefer-promise-reject-errors */
/* https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html */
export interface CancellablePromise<T> {
  promise: Promise<T>;
  cancel(): void;
}

const makeCancelable = <T>(promise: Promise<T>): CancellablePromise<T> => {
  let hasCanceled = false;

  const wrappedPromise = new Promise<T>((resolve, reject) => {
    promise.then(
      (val) => (hasCanceled ? reject({ isCanceled: true }) : resolve(val)),
      (error) => (hasCanceled ? reject({ isCanceled: true }) : reject(error)),
    );
  });

  return {
    promise: wrappedPromise,
    cancel(): void {
      hasCanceled = true;
    },
  };
};

export default makeCancelable;
