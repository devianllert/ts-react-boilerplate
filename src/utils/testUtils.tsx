/* eslint-disable import/prefer-default-export */
import React, { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { i18n as i18Next } from 'i18next';
import { I18nextProvider } from 'react-i18next';
import { render, RenderResult } from '@testing-library/react'; // eslint-disable-line

import configureStore, { EnhancedStore } from '../core/configureStore';
import i18nInstance from '../core/i18n';

export interface RenderProvidersOptions {
  initialState?: {};
  route?: string;
  routerHistory?: MemoryHistory;
  store?: EnhancedStore;
  i18n?: i18Next;
}

export interface RenderProvidersResult extends RenderResult {
  routerHistory: MemoryHistory;
  store: EnhancedStore;
  i18n: i18Next;
}

export type RenderProviders = (ui: ReactElement, options?: RenderProvidersOptions) => RenderProvidersResult;

const renderWithProviders: RenderProviders = (
  ui,
  {
    initialState = {},
    route = '/',
    routerHistory = createMemoryHistory({ initialEntries: [route] }),
    store = configureStore(initialState),
    i18n = i18nInstance,
  }: RenderProvidersOptions = {},
): RenderProvidersResult => ({
  ...render(
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        <Router history={routerHistory}>{ui}</Router>
      </I18nextProvider>
    </Provider>,
  ),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  store,
  routerHistory,
  i18n,
});

export {
  renderWithProviders,
};
