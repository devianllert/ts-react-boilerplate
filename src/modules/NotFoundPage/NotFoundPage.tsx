import React, { ReactElement } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { MdArrowBack } from 'react-icons/md';

import Button from '../../components/Button';
import Typography from '../../components/Typography';

import * as S from './styled';

const NotFoundPage = (): ReactElement => {
  const { t } = useTranslation();

  return (
    <>
      <Helmet>
        <title>{t('NOT_FOUND_TITLE')}</title>
      </Helmet>

      <S.CenteredContainer>
        <S.Image src="/img/not_found.svg" alt="not found" />

        <Typography variant="h1" align="center" gutterBottom>404</Typography>
        <Typography variant="h4" align="center" gutterBottom>{t('NOT_FOUND_TITLE')}</Typography>

        <Button to="/" startIcon={<MdArrowBack size={16} />}>{t('BACK')}</Button>
      </S.CenteredContainer>
    </>
  );
};

export default NotFoundPage;
