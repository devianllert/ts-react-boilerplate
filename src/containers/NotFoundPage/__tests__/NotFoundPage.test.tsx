import React from 'react';
import { cleanup } from '@testing-library/react';

import { renderWithRouter } from '../../../utils/testUtils';

import NotFoundPage from '../NotFoundPage';

import '@testing-library/jest-dom/extend-expect';

describe('<NotFoundPage />', (): void => {
  afterEach(cleanup);

  it('should render and match the snapshot', (): void => {
    const { container } = renderWithRouter(<NotFoundPage />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with initial state', (): void => {
    const { getByText } = renderWithRouter(<NotFoundPage />);

    expect(getByText(/404/)).toBeDefined();
  });
});
