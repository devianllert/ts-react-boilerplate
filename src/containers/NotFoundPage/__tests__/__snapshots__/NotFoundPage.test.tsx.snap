// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotFound /> should render and match the snapshot 1`] = `
<div
  class="content"
>
  <h1
    class="logo"
  >
    404
  </h1>
  <h2
    class="title"
  >
    notFound.title
  </h2>
  <span
    class="subtitle"
  >
    notFound.subtitle
  </span>
  <a
    class="button primary medium"
    href="/"
  >
    notFound.button
  </a>
</div>
`;
